### PR TITLE
Enhanced DeepSquareClient: Top-up Feature, Job Price Info, and Optional Wallet Support

### PR DESCRIPTION
**Pull request by efelem**
This update to DeepSquareClient includes three main features: an improved top-up functionality, access to detailed job pricing information, and optional wallet integration. These enhancements aim to provide users with a more direct and efficient experience when using DeepSquareClient.

- Added `@ethersproject/units` and `dayjs` dependencies to package.json
- Added `ProviderManager__factory` export in `src/contracts/factories/index.ts`
- Added `ProviderManager` export in `src/contracts/index.ts`
- Imported `dayjs` and `formatEther` from @ethersproject/units in src/index.ts
- Defined `JobStatus` enum, `isJobTerminated()`, `computeCost()`, and `computeCostPerMin()` functions
- Updated `DeepSquareClient` constructor with an optional privateKey
- Conditionally created and connected `Wallet`, `MetaScheduler`, `ProviderManager`, and `Credit` instances based on the presence of a valid privateKey
- Updated `submitJob()` and `cancel()` methods to throw an error if a valid privateKey is not provided
- Updated `getJob()` method to return additional information: `actualCost`, `costPerMin`, and `timeLeft`
- Added `topUp()` method to top up a job with specified amount of credits

### CODE REVIEW

**Comment by Darkness4** [comment]
I'm hijacking this branch and add similar changes to https://github.com/deepsquare-io/deepsquare-client/pull/9

File: src/index.ts [hunk 1]
```diff
@@ -1,75 +1,151 @@
...
+  SCHEDULED = 2,
+  RUNNING = 3,
+  CANCELLED = 4,
+  FINISHED = 5,
+  FAILED = 6,
+  OUT_OF_CREDITS = 7,
+}
 
-  private readonly graphqlClient: GraphQLClient;
+interface GetLogsMethodsReturnType {
```
Comment (Darkness4) [inline src/index.ts:41]: This is dead code (non exported, not used)

**Review by Darkness4 — CHANGES_REQUESTED** [review]

File: src/index.ts [hunk 2]
```diff
@@ -1,75 +1,151 @@
...
+    status === JobStatus.CANCELLED ||
+    status === JobStatus.FAILED ||
+    status === JobStatus.FINISHED ||
+    status === JobStatus.OUT_OF_CREDITS
+  );
+}
 
-  private readonly credit: Credit;
+// compute the job current cost (total if job finished)
+const computeCost = (job: any, provider: any): number => {
```
Comment (Darkness4) [inline src/index.ts:55]: Please find something else other than `any`.

File: src/index.ts [hunk 3]
```diff
@@ -1,75 +1,151 @@
...
+    ? +formatEther(job.cost.finalCost)
+    : +`${(
+        dayjs().diff(dayjs(job.time.start.toNumber() * 1000), 'minutes') *
+        computeCostPerMin(job, provider)
+      ).toFixed(0)}`;
+};
 
-  private readonly provider: JsonRpcProvider;
+// compute the job costs per minute based on provider price
+const computeCostPerMin = (job: any, provider: any): number => {
```
Comment (Darkness4) [inline src/index.ts:65]: Please find something else other than `any`.

Comment (Darkness4) [reply src/index.ts:65]: Also, prefer `function` over an anonymous function syntax.

File: src/index.ts (hunk 2 above)
Comment (Darkness4) [reply src/index.ts:55]: Also, prefer `function` over an anonymous function syntax.

**Review by Darkness4 — CHANGES_REQUESTED** [review]

**Review by Darkness4 — APPROVED** [review]
LGTM. @valentinpollart I'll let you have the final word.

File: src/grpc/client.ts [hunk 4]
```diff
@@ -1,11 +1,11 @@
...
-import { ChannelCredentials } from "@grpc/grpc-js";
 
 export function createLoggerClient(): LoggerAPIClient {
-        const transport = new GrpcTransport({
-                host: "grid-logger.deepsquare.run:443",
-                channelCredentials: ChannelCredentials.createSsl(),
-        });
-        return new LoggerAPIClient(transport);
+  const transport = new GrpcTransport({
+    host: "grid-logger.deepsquare.run:443",
```
Comment (efelem) [inline src/grpc/client.ts:7]: Shouldn't this be an environment variable ?

**Review by valentinpollart — APPROVED** [review]

File: src/grpc/client.ts (hunk 4 above)
Comment (Darkness4) [reply src/grpc/client.ts:7]: It's already a parameter, you can inject a factory function inside the `build` function.

**Pull request merged**